src/Coordination/KeeperStorage.cpp

```diff
@@ -226,12 +227,67 @@ void KeeperStorage::Node::shallowCopy(const KeeperStorage::Node & other)
     cached_digest = other.cached_digest;
 }
 
-KeeperStorage::KeeperStorage(int64_t tick_time_ms, const String & superdigest_, const bool digest_enabled_)
-    : session_expiry_queue(tick_time_ms), digest_enabled(digest_enabled_), superdigest(superdigest_)
+KeeperStorage::KeeperStorage(
+    int64_t tick_time_ms, const String & superdigest_, const KeeperContextPtr & keeper_context_, const bool initialize_system_nodes)
+    : session_expiry_queue(tick_time_ms), keeper_context(keeper_context_), superdigest(superdigest_)
 {
     Node root_node;
     container.insert("/", root_node);
-    nodes_digest += root_node.getDigest("/");
+    addDigest(root_node, "/");
+
+    if (initialize_system_nodes)
+        initializeSystemNodes();
+}
+
+void KeeperStorage::initializeSystemNodes()
+{
+    if (initialized)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "KeeperStorage system nodes initialized twice");
+
+    // insert root system path if it isn't already inserted
+    if (container.find(keeper_system_path) == container.end())
```

Comment:
How can it be already inserted if we do not load it from snapshot and `initializeSystemNodes` is called once?

Reply:
because we can insert empty root system nodes when loading from snapshot because of the consistency checks (`'/'.stat.numChildren == '/'.getChildren.size()`)